ws.1
.\" -*- nroff -*-
.\" Copyright (C) Ollie Etherington <www.etherington.xyz>
.\"
.\" This document is licensed under the terms of The GNU Free Documentation
.\" License, as published by the Free Software Foundation, version 1.2 or
.\" (at your option) any later version, with no Invariant Sections, no
.\" Front-Cover Texts, and no Back-Cover Texts.  You should have received a
.\" copy of the GNU Free Documentation License along with this program.  If
.\" not, see <http://www.gnu.org/licenses/>.
.\"
.\" Update the date below whenever revising:
.TH WS 1 "6 Jun 2021" "Version 1.0.0"

.SH NAME
ws - A lightweight workspace manager

.SH SYNOPSIS
.B ws
.I [OPTIONS]\ [WORKSPACE NAME]
.br

.SH DESCRIPTION
This manual page briefly documents the \fBws\fP command.
.PP
\fBws\fP is a free lightweight workspace manager for use on systems running
Bash with configuration for editor invocation and support for multiple
workspace definitions per directory.
.PP
Global configuration is done in the \fIwsrc\fP file described in detail in it's
own manpage; \fBwsrc\fP(5).
.PP
Workspaces are defined in directory specific files called \fI.ws\fP. Their
syntax is described below in the section entitled \fIwriting a .ws file\fP.

.SH OPTIONS
.TP
.B \-e (\-\-edit)
Edit (or create) the .ws configuration file for the current directory.
.TP
.B \-p (\-\-print)
Instead of running the final editor command, print it to stdout instead.
.TP
.B \-l (\-\-list)
List the available wordspaces and the files they contain.
.TP
.B \-h (\-\-help)
Display a help message specifying command line usage.
.TP
.B \-v (\-\-version)
Display program version information.

.SH WRITING A .WS FILE
Workspaces a defined at the directory level in a file called \fI.ws\fP.
Invoking \fBws\fP in a directory without providing a workspace name on the
command line will open the default workspace which is, by default, called
\fBws\fP. It is defined like so:
.PP
\fBws { MY_FILE_1 MY_FILE_2 }\fP
.PP
Note that whitespace is syntactically unimportant - this definition could have
been split over several lines for easier reading and editing.
.PP
In the event that multiple workspaces in the same directory are given the same
name, the FIRST definition will always be launched.
.PP
Normally workspaces use the default editor command which can be set in the
wsrc configuration file but it's also possible to give workspaces their own
custom editor commands. This is done with the following syntax:
.PP
\fBws - "nano --mouse" { MY_FILE_1 }\fP
.PP
This will invoke the command "nano --mouse MY_FILE_1". Note that the whitespace
either side of the hyphen is necessary otherwise it will be interpreted as a
workspace name with a dash in it. Single word commands such as "nano" or
"emacs" don't require quotes but any command with whitespace in does to avoid
ambiguity.

.SH BUGS
Please send any comments or bug reports to \fBhttps://github.com/oetherington/ws\fP.

.SH HOMEPAGE
www.etherington.xyz

.SH SEE ALSO
.PD 0
.TP
\fBwsrc\fP(5)

.SH AUTHOR
Ollie Etherington <www.etherington.xyz>
.PP
This document is available under the terms of the GNU Free Documentation
License.


ws.sh
#!/bin/bash

# ws workspace manager
# Free software under the GNU GPLv3

# Info
NAME="WS"
DESCRIPTION="CLI Workspace Manager"
VERSION="1.0.0"
COPYRIGHT="Copyright Ollie Etherington"
WEBSITE="www.etherington.xyz"
LICENSE="Free software under the GNU GPLv3"

# Program data
RUNCMD=${EDITOR:="vim -p"}	# Get the EDITOR env var if it exists, else set it
RC="$HOME/.wsrc"	# Global RC config
CFG=".ws"			# Local definitions#tokens=( `cat ${CFG} | tr ' ' '\n'` )
DEFAULT_WS="ws"		# The default workspace

# Check for system config and process it if it exists
# Setup the internal field seperator for parsing
backupIFS=$IFS
IFS="="

if [ -f $RC ]; then
	while read -r rcvar value; do
		case $rcvar in
		editor)
			RUNCMD="${value//\"/}"
			;;
		default_ws)
			DEFAULT_WS="${value//\"/}"
			;;
		*)
			echo "Warning: Unrecognized config option '$rcvar' in '$RC'"
			;;
		esac
	done < $RC
fi

# The workspace to run
WS=$DEFAULT_WS

# Restore the internal field seperator for later
IFS=$backupIFS

# Parse command line arguments
echo_cmd=false
list_ws=false
using_custom_ws_name=false
argv=( $@ )
argc=$#

# Help message
function print_help {
	echo $NAME - $DESCRIPTION
	echo $WEBSITE
	echo "Usage: ws [WORKSPACE NAME] [OPTIONS]..."
	echo Save common file groups to directory specific workspaces and associate
	echo them with an editor command
	echo The following options may be used:
	echo "  -h,--help     Show this help message"
	echo "  -v,--version  Show version information"
	echo "  -e,--edit     Edit the current directories .ws file"
	echo "  -p,--print    Instead of executing the run command, print it"
	echo "  -l,--list     List the available workspaces"
	exit 0
}

# Version info
function print_version {
	echo $NAME - $DESCRIPTION
	echo Version $VERSION
	echo $WEBSITE
	echo $COPYRIGHT
	echo $LICENSE
	exit 0
}

# Edit this directories workspace definition file
function edit_spec {
	$RUNCMD $CFG
	exit 0
}

for i in `seq 0 $(($argc-1))`; do
	case "${argv[i]}" in
	-h|--help)
		print_help
		;;
	-v|--version)
		print_version
		;;
	-e|--edit)
		edit_spec
		;;
	-p|--print)
		echo_cmd=true
		;;
	-l|--list)
		list_ws=true
		;;
	*)
		if $using_custom_ws_name; then
			echo "Error: Multiple workspaces given ('$WS' and '${argv[i]}')"
			exit 0
		else
			WS=${argv[i]}
			using_custom_ws_name=true
		fi
		;;
	esac
done

# Ensure the workspace config file exists
if [ ! -f $CFG ]; then
	echo Workspace config file not found - exiting
	exit 0
fi

# Get workspace definition
definition=`cat ${CFG}`
tokens=( $definition )
num_tokens=${#tokens[@]}
max_token=$(($num_tokens-1))
ws_not_found=true
file_hook=""

# If we're listing the workspaces, do that and then exit
if [ $list_ws == true ]; then
	i=0
	while [ $i -lt $max_token ]; do
		# Get the workspace name
		WS=${tokens[i]}

		((i++))

		# Check is we're using a custom editor command
		if [ ${tokens[i]} == "-" ]; then
			((i++))

			EDITOR_CMD=${tokens[i]}

			# If it's multiple tokens in quotes we need to get the rest
			if [[ $EDITOR_CMD == \"* ]]; then
				while [[ $EDITOR_CMD != *\" ]]; do
					((i++))
					EDITOR_CMD="$EDITOR_CMD ${tokens[i]}"
				done
			fi

			# Remove the actual quotes from the string
			EDITOR_CMD=`echo $EDITOR_CMD | sed 's/^"\(.*\)"$/\1/'`

			((i++))
		else
			EDITOR_CMD=$RUNCMD
		fi

		# Print the name and editor command
		echo "$WS (Command: $EDITOR_CMD)"

		# Make sure we have a list of files
		if [ ${tokens[i]} != "{" ]; then
			echo "Syntax Error: Expected '{' found '${tokens[i]}'"
			exit 0
		fi

		((i++))

		# We do - get them
		while [ ${tokens[i]} != "}" ]; do
			# Make sure we're still in bounds
			if [ $i == $max_token ]; then
				echo "Syntax Error: Expected '}' found EOF"
				exit 0
			fi

			echo "  - ${tokens[i]}"

			((i++))
		done

		((i++))
	done

	exit 0
fi

# Do the parsing
for i in `seq 0 $max_token`; do
	# Only bother properly parsing the workspace we actually want to open
	if [ ${tokens[i]} == $WS ]; then
		ws_not_found=false

		((i++))

		# Check if this workspace has a custom editor command
		if [ ${tokens[i]} == "-" ]; then
			((i++))

			# It does - get it
			NEWCMD=${tokens[i]}

			# If it's multiple tokens in quotes we need to get the rest
			if [[ $NEWCMD == \"* ]]; then
				while [[ $NEWCMD != *\" ]]; do
					((i++))
					NEWCMD="$NEWCMD ${tokens[i]}"
				done
			fi

			# Remove the actual quotes from the string
			NEWCMD=`echo $NEWCMD | sed 's/^"\(.*\)"$/\1/'`

			# Set the editor command
			RUNCMD=$NEWCMD

			((i++))
		fi

		# Make sure we have a list of files
		if [ ${tokens[i]} != "{" ]; then
			echo "Syntax Error: Expected '{' found '${tokens[i]}'"
			exit 0
		fi

		((i++))

		# We do - get them
		while [ ${tokens[i]} != "}" ]; do
			# Make sure we're still in bounds
			if [ $i == $max_token ]; then
				echo "Syntax Error: Expected '}' found EOF"
				exit 0
			fi

			file_hook="$file_hook ${tokens[i]}"

			((i++))
		done
	fi
done

# Check if we didn't find the workspace
if $ws_not_found; then
	echo "Workspace '$WS' not found - exiting"
	exit 0
fi

# Start the workspace (or echo it...)
if $echo_cmd; then
	echo $RUNCMD $file_hook
else
	$RUNCMD $file_hook
fi
